wandb/internal/tb_watcher.py

```diff
@@ -0,0 +1,279 @@
+    if base_path is None:
+        base_path = os.path.dirname(path)
+    files_dir = sender._settings.files_dir
+    file_name = os.path.relpath(path, base_path)
+    abs_path = os.path.abspath(path)
+    wandb_path = os.path.join(files_dir, file_name)
+    util.mkdir_exists_ok(os.path.dirname(wandb_path))
+    # We overwrite existing symlinks because namespaces can change in Tensorboard
+    if os.path.islink(wandb_path) and abs_path != os.readlink(wandb_path):
+        os.remove(wandb_path)
+        os.symlink(abs_path, wandb_path)
+    elif not os.path.exists(wandb_path):
+        os.symlink(abs_path, wandb_path)
+    # TODO(jhr): need to figure out policy, live/throttled?
+    sender._save_file(file_name)
+
+
+class TBWatcher(object):
+    def __init__(self, settings, sender=None):
+        self._logdirs = {}
+        self._consumer = None
+        self._settings = settings
+        self._sender = sender
+        # TODO(jhr): do we need locking in this queue?
+        self._watcher_queue = queue.PriorityQueue()
+
+    def _calculate_namespace(self, logdir):
+        dirs = list(self._logdirs) + [logdir]
+        rootdir = os.path.dirname(os.path.commonprefix(dirs))
+        if os.path.isfile(logdir):
+            filename = os.path.basename(logdir)
+        else:
+            filename = ""
+        # Tensorboard loads all tfevents files in a directory and prepends
+        # their values with the path.  Passing namespace to log allows us
+        # to nest the values in wandb
+        namespace = logdir.replace(filename, "").replace(rootdir, "").strip(os.sep)
+        # TODO: revisit this heuristic, it exists because we don't know the
+        # root log directory until more than one tfevents file is written to
+        if len(dirs) == 1 and namespace not in ["train", "validation"]:
+            namespace = None
+        return namespace
+
+    def add(self, logdir, save):
+        if logdir in self._logdirs:
+            return
+        namespace = self._calculate_namespace(logdir)
+        # TODO(jhr): implement the deferred tbdirwatcher to find namespace
+
+        if not self._consumer:
+            self._consumer = TBEventConsumer(self, self._watcher_queue)
+            self._consumer.start()
+
+        tbdir_watcher = TBDirWatcher(self, logdir, save, namespace, self._watcher_queue)
+        self._logdirs[logdir] = tbdir_watcher
+        tbdir_watcher.start()
+
+    def finish(self):
+        for tbdirwatcher in six.itervalues(self._logdirs):
+            tbdirwatcher.shutdown()
+        for tbdirwatcher in six.itervalues(self._logdirs):
+            tbdirwatcher.finish()
+        if self._consumer:
+            self._consumer.finish()
+
+
+class TBDirWatcher(object):
+    from tensorboard.backend.event_processing import directory_watcher  # type: ignore
+    from tensorboard.backend.event_processing import event_file_loader  # type: ignore
+    from tensorboard.compat import tf as tf_compat  # type: ignore
```

Comment:
I'm a little worried about us not having tensorboard and how we handle the internal process crashing.  We should probably make an internal exception class that the process loop catches and returns an error object to the user process.

Also, wild idea, instead of doing the insane PTY console logic, we could have the user process queue up log messages from the backend process and flush them when any method on the run object is called...

Reply:
good catch, i thought imports inside classes wouldnt get called unless instantiated but i was confused.
converted to get_module.  

I agree with the internal process robustness need, I will look into that in a later PR.  At this point we shouldnt ever blow up in internal because we won't initialize this object unless the patch succeeded